server/src/main/java/org/elasticsearch/common/util/concurrent/KeyedLock.java

```diff
@@ -48,55 +50,36 @@ public KeyedLock() {
      * by the same thread multiple times. The lock is released by closing the returned {@link Releasable}.
      */
     public Releasable acquire(T key) {
-        while (true) {
-            KeyLock perNodeLock = map.get(key);
-            if (perNodeLock == null) {
-                ReleasableLock newLock = tryCreateNewLock(key);
-                if (newLock != null) {
-                    return newLock;
-                }
-            } else {
-                assert perNodeLock != null;
-                int i = perNodeLock.count.get();
-                if (i > 0 && perNodeLock.count.compareAndSet(i, i + 1)) {
-                    perNodeLock.lock();
-                    return new ReleasableLock(key, perNodeLock);
-                }
-            }
-        }
+        KeyLock perNodeLock = map.compute(key, computeLock(fair));
```

Comment:
> I wonder if optimistically doing tryCreateNewLock before any map lookup would not be faster, since in the uncontended case that would go through and in the contended case, it does not matter?

I guess that may be true. I'm just a little worried about how bad the contended case will behave then. It seems like in the real world, most of the time gained by my change comes from looking up the right bucket in the map only once. In the end it's really hard to measure the effect of this change in the contended scenario well. For the uncontended case, this solution wins hands down in some quick benchmarks just by only doing the map lookup once and inlining much better. For contended case I wasn't able to find a good benchmark approximation because it depends massively on the specific keys used etc. I would hope though that on average, that spinning on a `synchronized` block (I think no matter what the lock will rarely be contended to the point where it's not a spin-lock) is still faster than spinning by doing object creation and repeated map lookups.

Reply:
Can you share (perhaps privately) the benchmarks done with and without the change and the results?

If we have done benchmarks, showing this is faster in the uncontended case, then I am good with this.